internal/client/consensus/grandpa/environment.go

```diff
@@ -625,3 +630,1018 @@ type votePrecommit[H runtime.Hash, N runtime.Number] struct {
+) {
+	logger.Warnf("Detected precommit equivocation in the finality worker: %v", equivocation)
+	err := e.reportEquivocation(primitives.EquivocationPrecommit[H, N](equivocation))
+	if err != nil {
+		logger.Warnf("Error reporting precommit equivocation: %s", err)
+	}
+}
+
+func bestChainContaining[
+	H runtime.Hash,
+	N runtime.Number,
+	Hasher runtime.Hasher[H],
+	Header runtime.Header[N, H],
+	E runtime.Extrinsic,
+](
+	block H,
+	client ClientForGrandpa[H, N, Hasher, Header, E],
+	authoritySet *SharedAuthoritySet[H, N],
+	selectChain common.SelectChain[H, N, Header],
+	votingRule VotingRule[H, N, Header],
+) (value *grandpa.HashNumber[H, N], err error) {
+
+	var baseHeader Header
+	h, err := client.Header(block)
+	if err != nil {
+		return nil, err
+	}
+	if h == nil {
+		logger.Warnf("Encountered error finding best chain containing %s: couldn't find base block: %s", block, err)
+		return nil, nil
+	}
+	baseHeader = *h
+
+	// We refuse to vote beyond the current limit number where transitions are scheduled to occur.
+	// Once blocks are finalized that make that transition irrelevant or activate it, we will
+	// proceed onwards. most of the time there will be no pending transition.  The limit, if any, is
+	// guaranteed to be higher than or equal to the given base number.
+	limit := authoritySet.currentLimit(baseHeader.Number())
+	logger.Debugf("Finding best chain containing block %s with number limit %s", block, limit)
+
+	var targetHeader Header
+	th := <-selectChain.FinalityTarget(block, nil)
+	if th.Error != nil {
+		logger.Debugf(
+			"Encountered error finding best chain containing %s: couldn't find target block: %s", block, th.Error)
+		// NOTE: in case the given SelectChain doesn't provide any block we fallback to using the given base block
+		// provided by the GRANDPA voter.
+		//
+		// For example, LongestChain will error if the given block to use as base isn't part of the best chain
+		// (as defined by LongestChain), which could happen if there was a re-org.
+		targetHeader = baseHeader
+	} else {
+		h, err := client.Header(th.Hash)
+		if err != nil {
+			return nil, err
+		}
+		if h == nil {
+			panic("Header known to exist after FinalityTarget call.")
+		}
+		targetHeader = *h
+	}
+
+	// NOTE: this is purposefully done after FinalityTarget to prevent a case where in-between these two requests
+	// there is a block import and FinalityTarget returns something higher than BestChain.
+	var bestHeader Header
+	bh := <-selectChain.BestChain()
+	if bh.Error != nil {
+		logger.Warnf(
+			"Encountered error finding best chain containing %s: couldn't find best block: %s", block, bh.Error)
+		return nil, nil //nolint: nilerr
+	}
+	bestHeader = bh.Header
+
+	isDescendentOf := utils.IsDescendantOf[H, N, Header](client, nil)
+
+	if targetHeader.Number() > bestHeader.Number() ||
+		targetHeader.Number() == bestHeader.Number() && targetHeader.Hash() != bestHeader.Hash() {
```

Comment:
```suggestion
	if targetHeader.Number() >= bestHeader.Number() && targetHeader.Hash() != bestHeader.Hash() {
```

Reply:
Shouldnt the decendant check be here as well? https://github.com/paritytech/polkadot-sdk/blob/153738c2baef5fb7a211f2a4d6462a0d1108fff7/substrate/client/consensus/grandpa/src/environment.rs#L1256

Reply:
I implemented it wrong, I've cleaned it up in 2d9409bc65ca2f14fd44fe161872a4bdce7492b3